[Feat] 대학교 리스트 검색 도메인 & 네트워킹

tea-hkim opened the pull request:
## 작업내용
<!-- 작업 내용과 이미지를 첨부해주세요. -->
- [x] 대학교 리스트 fetch
- [x] 검색 버튼과 모델 연결


<!-- (+스크린샷)이 있다면 적어주세요. 없으면 지워주세요-->
|'서울' 검색|'애플'검색|
|:---:|:---:|
|<img width="250" src="https://user-images.githubusercontent.com/81206228/200205299-ea2f6248-3ed6-4037-a453-addb045ff4a0.png">|<img width="250" src="https://user-images.githubusercontent.com/81206228/200205339-0608abe3-b124-4051-aa91-2da2aa299c1b.png">|

## 리뷰포인트
<!-- 리뷰가 필요한 포인트와 해당 되는 커밋을 링크로 걸어주세요. -->
### ViewModel의 역할에 대한 고민
 - 로직을 변경한 이유는 cell 내부의 UI component들에 대한 접근 권한을 private 하게 관리하기 위함입니다.
   변경 전 코드에는 cell 내부의 UILabel의 text에 직접 전달을 했습니다.
   하지만 bind라는 함수를 사용해서 외부에서 데이터를 받아와 UILabel을 그려주는 방식으로 변경하였습니다.
 - searchedOrgArray라는 변수명의 역할을 명확하게 하기 위해 orgNameArrary에서 필터링하는 방식이 아닌 orgArray에서 바로 필터링하는 방식으로 변경했습니다.
 - 또한 다음 뷰에서는 organization의 이름과 도메인이 필요하기 때문에 직접 organization 데이터를 전달하는 편이 효율적이라고 생각되어 변경했습니다.

 고민인 부분
 - View와 직접적인 관계가 있는 데이터를 가지고 있는 것이 ViewModel의 역할이라고 생각합니다.
   하지만 제가 구현한 방식은 OrganizatinoListView에서 원하는 데이터인 조직 이름 리스트가 아니라 조직의 리스트를 가지고 있고 이 방식이 과연 맞는 방식인지에 대한 의문이 남았습니다.

   전에 구현하던 방식(조직 이름 리스트만 가지고 있는 방식)이 ViewModel의 역할을 제대로 수행하고 있다고 생각합니다. 
View에 띄워야 하는 데이터만을 가지고 있기 때문입니다. 
하지만 이 방식에서 문제라고 생각되는 부분은 cell이 탭되어 다른 뷰로 넘어갔을 경우 ViewModel에서 orgArray에서 해당 조직의 이름과 같은 organization을 찾아 넘겨줘야 한다는 부분에서 비효율적이라는 생각이 들었습니다. 
이 경우에는 이런 역할을 수행하는 함수를 따로 만들어서 cell이 탭이 되어 뷰가 전환될 때 함수를 실행해줘야 할까요?


## Checklist
- [x] 브랜치를 가져와 작업한 경우 이전 브랜치에 PR을 보냈는지 확인
- [x] 빌드를 위해 SceneDelegate 수정한 것 PR로 올리지 않았는지 확인
- [x] 필요없는 주석, 프린트문 제거했는지 확인
- [x] 컨벤션 지켰는지 확인
- [x] final, private 제대로 넣었는지 확인
- [x] 다양한 디바이스에 레이아웃이 대응되는지 확인
  - [x] iPhone SE
  - [x] iPhone 13
  - [x] iPhone 13 Pro Max

tea-hkim committed:
Chore:broom:: OrganizationUseCase 변경

 - 내부에 publisher를 두는 것 보다 반환값으로 api에서 온 publisher를 바로 전달하는 것이 효율적일 것 같아서 변경하였습니다.
 - return 값으로 API의 결과물을 바로 전달하도록 하였습니다

tea-hkim committed:
Feat:sparkles:: OrganizatinoListViewModel과 OrganizatinoListUseCase 연결

 -  OrganizatinoListUseCase에서 fetch를 해오는 결과물을 바로 내부의 배열과 연결하였습니다

 결과
 1. orgArray에 organization의 타입을 가진 배열 생성
 2. orgNameArray에 API에서 dto로 변환된 대학교 이름만 가진 배열 생성

 추후 작업할 내용
 1. orgNameArray를 따로 만드는 것은 비효율적이라고 생각해 cell에 orgArray를 바로 전달할 예정입니다.
 2. 이렇게 작업하는 이유는 이 다음 뷰에서 organization의 도메인이 필요하기 때문입니다.
    그래서 cell이 가지고 있는 organization 모델을 바로 전달해서 뷰를 쉽게 그리도록 하기 위함입니다.

tea-hkim committed:
Merge branch 'dev' of https://github.com/DeveloperAcademy-POSTECH/MacC-TEAM-ZESTY into feat/OrgSettin/ZES-162-org-list-domain-setting

tea-hkim committed:
Fix:ladybug:: SearchTextField의 placeholder 설정

 - placeholder를 설정했습니다.
 - 예전에 대학교 검색 뷰 UI 작업에서 placeholder를 설정했는데 dev를 pull 해온 결과 placeholder가 사라진 것 같습니다

tea-hkim committed:
Chore:broom:: setInitialSearchedArray 삭제

 - api를 연결하기 전에 남아있던 코드여서 삭제했습니다

tea-hkim committed:
Refactor:recycle:: searchedOrgArray 로직 변경

 - 로직을 변경한 이유는 cell 내부의 UI component들에 대한 접근 권한을 private 하게 관리하기 위함입니다.
   변경 전 코드에는 cell 내부의 UILabel의 text에 직접 전달을 했습니다.
   하지만 bind라는 함수를 사용해서 외부에서 데이터를 받아와 UILabel을 그려주는 방식으로 변경하였습니다.
 - searchedOrgArray라는 변수명의 역할을 명확하게 하기 위해 orgNameArrary에서 필터링하는 방식이 아닌 orgArray에서 바로 필터링하는 방식으로 변경했습니다.
 - 또한 다음 뷰에서는 organization의 이름과 도메인이 필요하기 때문에 직접 organization 데이터를 전달하는 편이 효율적이라고 생각되어 변경했습니다.

 의문점
 - ViewModel은 View에 띄워지는 데이터를 가지고 있어야 한다고 생각합니다. View와 직접적인 관계가 있는 데이터를 가지고 있는 것이 ViewModel의 역할이라고 생각합니다.
   하지만 제가 구현한 방식은 OrganizatinoListView에서 원하는 데이터인 조직 이름 리스트가 아니라 조직의 리스트를 가지고 있고 이를 View로 전달해서 View에서 cell에게 조직을,
   다음 뷰에게도 조직을 전달하는 방식이 과연 맞는 방식인지에 대한 의문이 남았습니다.
   전에 구현하는 방식, 즉 조직 이름 리스트만 가지고 있는 방식이 ViewModel의 역할을 제대로 수행하고 있다고 생각합니다.
   하지만 이 방식의 문제라고 생각되는 부분은 cell이 탭되어 다른 뷰로 넘어갔을 경우 ViewModel에서 orgArray에서 해당 조직의 이름과 같은 organization을 찾아 넘겨줘야 한다는 부분에서 비효율적이라는 생각이 들었습니다.
   어떤 방식이 더 옳은 방식일까요?

tea-hkim committed:
Feat:sparkles:: TableView delegate 연결

 - delegate의 위치를 configureUI로 변경하였습니다.
 - cell이 탭 되었을 경우의 delegate를 추가하였습니다.
 - cell이 탭 된 경우 domainSettingViewController가 열리도록 수정하였습니다.

tea-hkim committed:
Feat:sparkles:: searchButton에 검색 기능 추가

 - ViewModel에 subject를 추가하여 검색버튼이 눌린 경우 검색 기능이 작동하도록 구현하였습니다.

tea-hkim committed:
Chore:broom:: 줄바꿈 제거

tea-hkim committed:
Feat:sparkles:: DomainSettingUseCase 생성

 - 기존 UserSignupUseCase에 있던 postSingupUser를 삭제하였습니다.
 - DomainSettingUseCase를 만들어 postSingupUser를 postUserEmail로 변경하였습니다
 - API문서 그대로 postUserEmail의 경우 user api 영역에 있으므로 UserAPI에 남겨뒀습니다.
   하지만 프론트엔드에서는 domainSetting의 영역이기 때문에 따로 DomainSettingUseCase를 만들어 UserAPI에서 호출하였습니다

tea-hkim committed:
Feat:sparkles:: DomainSettingUseCase 연결

 - DomainSettingViewController와 DomainSettingUseCase를 연결했습니다
 - DomainSettingViewController와 OrganizationListViewController를 연결했습니다
   pushViewController를 통해 넘어오기 전, DomainSettingViewController의 파라미터 값으로 가지고 있던 Organization을 넘겨주도록 하였습니다

Gobans commented on Projects/App/Sources/UI/Organization/OrganizationList/ViewModel/OrganizatinoListViewModel.swift:
```diff
@@ -11,58 +11,92 @@ import Foundation
     private var cancelBag = Set<AnyCancellable>()
     
     // Input
     @Published var userTextInput: String = ""
+    let isSearchText = PassthroughSubject<String, Never>()
```
p1
요거 불필요한 중복 바인딩인 것 같습니다.
userTextInput 하나만 사용하고 searchButtonTapped 했을 떄
` searchingTextField.resignFirstResponder()` 하는게 동작이 더 자연스러운 것 같네요

tea-hkim replied:
이 방법이 있었네요!
4601a6200444efd19800947dd41f0c591b726b6a 변경했습니다!

Gobans reviewed (changes requested):
고생하셨습니다 닉!
viewModel에서 필요없는 데이터를 useCase 에서 받아 온다면, 로직을 useCase 로 옮기는게 좋을 것 같습니다.
그리고 useCase 에서 데이터를 처리&가공을 해서 viewModel이 가지고 있어야하는 데이터를 내려주면 될 것 같네요

tea-hkim committed:
Chore:broom:: print 제거

tea-hkim committed:
Chore:broom:: authentication todo 설정

tea-hkim committed:
Chore:broom:: DomainSettingViewModel 내부 Organization 접근권한 변경

 - private을 삭제헀습니다

tea-hkim committed:
Feat:sparkles:: viewWillAppear 추가

 - VerifingCodeViewController로 뷰가 전환되고 나서 다시 현재의 뷰로 돌아올 경우 키보드가 내려가는 현상이 있었습니다
 - 해당 문제를 viewWillAppear 내부에 뷰에 들어오자마자 키보드가 올라오는 코드를 추가하여 해결했습니다.

tea-hkim committed:
Chore:broom:: ViewModel isInputValid 로직 변경

 - 기존의 직관적이지 않은 논리값에서 변수명에 직관적인 논리값 전환으로 변경하였습니다

tea-hkim committed:
Chore:broom:: searchedButton 중복된 로직 수정

tea-hkim committed:
Feat:sparkles:: 뷰에 들어오면 키보드 활성화

tea-hkim committed:
Merge pull request #98 from DeveloperAcademy-POSTECH/feat/OrgSetting/ZES-163-org-domain-setting

[WIP][Feat] 대학교 이메일 입력 도메인 & 네트워킹

chaneeii commented on Projects/App/Sources/UI/Organization/DomainSetting/ViewModel/DomainSettingViewModel.swift:
```diff
@@ -10,32 +10,63 @@ import Combine
+            }
+            .store(in: &cancelBag)
+        
+        useCase.isEmailOverlapedSubject
+            .sink { [weak self] isEmailOverlaped in
+                guard let self = self else { return }
+                print(#function)
+                if isEmailOverlaped {
+                    self.shouldDisplayWarning = true
+                }
+                self.isEmailOverlapedSubject.send(isEmailOverlaped)
+                
+            }
+            .store(in: &cancelBag)
```
p5)
init 이 너무 길어져서 함수를 따로 파도 좋을 것 같아요!

chaneeii reviewed (approved):
고생하셨습니다..!
이제 컴바인 사용이 왜죠 엄청 잘쓰시는거같아 보이는거같ㅇ요...!🥳

tea-hkim committed:
Merge branch 'dev' into feat/OrgSettin/ZES-162-org-list-domain-setting
